File: solutions/2026-09-26/java/2026-09-26-evaluate-the-bracket-pairs-of-a-string.java
/**
 * Problem: Evaluate the Bracket Pairs of a String
 * Link: https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string/
 *
 * Approach:
 * 1. Store the key-value pairs from the `knowledge` array into a HashMap for efficient lookups.
 * 2. Iterate through the input string `s`.
 * 3. When an opening parenthesis '(' is encountered, it signifies the start of a bracket pair.
 * 4. Extract the key within the parentheses.
 * 5. Look up the key in the HashMap.
 * 6. If the key is found, append its corresponding value to the result string.
 * 7. If the key is not found, append "?" to the result string.
 * 8. If a character is not part of a bracket pair (i.e., not an opening parenthesis), append it directly to the result string.
 * 9. Use a StringBuilder for efficient string concatenation.
 *
 * Time Complexity: O(N + K), where N is the length of the string `s` and K is the number of entries in `knowledge`.
 *                  The HashMap creation takes O(K) time. Iterating through `s` takes O(N) time, with each lookup in the HashMap being O(1) on average.
 *
 * Space Complexity: O(K), where K is the number of entries in `knowledge`, to store the key-value pairs in the HashMap.
 */
import java.util.HashMap;
import java.util.Map;

class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        // Create a HashMap to store the knowledge for efficient lookups.
        // The key of the map will be the bracketed key, and the value will be its replacement.
        Map<String, String> knowledgeMap = new HashMap<>();
        for (List<String> pair : knowledge) {
            knowledgeMap.put(pair.get(0), pair.get(1));
        }

        // Use a StringBuilder to efficiently build the result string.
        StringBuilder result = new StringBuilder();
        // `i` is the current index in the input string `s`.
        int i = 0;
        // Iterate through the input string `s`.
        while (i < s.length()) {
            // Check if the current character is an opening parenthesis, indicating the start of a bracket pair.
            if (s.charAt(i) == '(') {
                // Move `i` to the character after the opening parenthesis.
                i++;
                // Find the index of the closing parenthesis.
                int closingBracketIndex = s.indexOf(')', i);
                // Extract the key from the bracket pair.
                String key = s.substring(i, closingBracketIndex);

                // Look up the key in the knowledgeMap.
                // If the key exists, append its value.
                // If the key does not exist, append "?".
                result.append(knowledgeMap.getOrDefault(key, "?"));

                // Move `i` to the character after the closing parenthesis.
                i = closingBracketIndex + 1;
            } else {
                // If the current character is not an opening parenthesis, it's a regular character.
                // Append it directly to the result string.
                result.append(s.charAt(i));
                // Move to the next character.
                i++;
            }
        }

        // Convert the StringBuilder to a String and return it.
        return result.toString();
    }
}

File: solutions/2026-09-26/python/2026-09-26-evaluate-the-bracket-pairs-of-a-string.py
```python
# Solves the "Evaluate the Bracket Pairs of a String" LeetCode problem.
# Problem Link: https://leetcode.com/problems/evaluate-the-bracket-pairs-of-a-string/
#
# Approach:
# The problem requires replacing bracketed key-value pairs with their corresponding values from a knowledge base.
# If a key is not found in the knowledge base, it should be replaced with '?'.
# We can use a hash map (dictionary in Python) to store the knowledge base for efficient lookups.
# We iterate through the input string `s`. When we encounter an opening bracket '(', we start collecting the characters
# until we find a closing bracket ')'. This collected string is the key.
# We then look up this key in our knowledge map.
# If the key exists, we append its corresponding value to our result.
# If the key does not exist, we append '?' to our result.
# If we encounter a character that is not part of a bracket pair (i.e., not '(' or ')'), we simply append it to our result.
#
# Time Complexity: O(N + K), where N is the length of the string `s` and K is the number of entries in `knowledge`.
#   - Building the knowledge map takes O(K) time.
#   - Iterating through the string `s` to build the result string takes O(N) time in the worst case.
#   - String concatenation can take O(length of string), but by using a list and then joining, it becomes amortized O(N).
# Space Complexity: O(K + N), where K is the number of entries in `knowledge` and N is the length of the output string.
#   - The knowledge map stores up to K key-value pairs.
#   - The result string can be up to O(N) in length.

class Solution:
    def evaluate(self, s: str, knowledge: list[list[str]]) -> str:
        """
        Evaluates bracket pairs in a string based on provided knowledge.

        Args:
            s: The input string with bracket pairs.
            knowledge: A list of [key, value] pairs.

        Returns:
            The string with evaluated bracket pairs.
        """

        # Create a dictionary to store the knowledge for efficient lookups.
        # This maps keys to their corresponding values.
        knowledge_map = {key: value for key, value in knowledge}

        # Use a list to build the result string efficiently.
        # Appending to a list and then joining is generally faster than repeated string concatenation.
        result_parts = []

        # Pointer to iterate through the string `s`.
        i = 0
        n = len(s)

        # Iterate through the string `s`.
        while i < n:
            # If the current character is an opening bracket '(':
            if s[i] == '(':
                # Find the index of the corresponding closing bracket ')'.
                j = i + 1
                while j < n and s[j] != ')':
                    j += 1

                # Extract the key from the bracket pair.
                # The key is the substring between '(' and ')'.
                key = s[i+1:j]

                # Look up the key in the knowledge map.
                # If the key is found, append its value to result_parts.
                # If the key is not found, append '?' to result_parts.
                result_parts.append(knowledge_map.get(key, "?"))

                # Move the pointer `i` to the position after the closing bracket.
                i = j + 1
            else:
                # If the current character is not an opening bracket,
                # it's a regular character that should be appended directly.
                result_parts.append(s[i])
                # Move the pointer `i` to the next character.
                i += 1

        # Join all the parts in the result_parts list to form the final string.
        return "".join(result_parts)

```